### DISCUSSION TITLE
The provider AutoDisposeFutureProvider was disposed before a value was emitted.

### DISCUSSION THREAD

**Discussion by perret123** [rrousselGit/riverpod#1343] [Q&A] [answered]
Hi! 

I use `flutter_riverpod 1.0.3 ` and I wonder if I am doing things correctly:

I have the following case: I use a StreamProvider to watch the results of a Firestore query. This look as such:

```
final guestPromosStreamProvider = StreamProvider.autoDispose.family<List<GuestPromo>, String?>((ref, guestId) {
  final database = ref.watch(databaseProvider);
  final businessId = ref.watch(businessIdProvider);
  ref.maintainState = true;
  return businessId != null && guestId != null ? database.guestPromosStream(businessId, guestId) : const Stream.empty();
});
```

This works great when I use it on a screen to display "live" information with `ref.watch` on this StreamProvider; all new guestPromo's appear on screen as soon as they are added in Firestore.

Now, I also want to be able to get the latest version of these guestPromos on a button tap - from a screen that doesn't need to load these guestPromos in advance - it only needs them after the button is pressed.

So my idea was to use a FutureProvider to get the .future of the StreamProvider to get the current result. The Future Provider looks like this:

```
final guestPromosProvider = FutureProvider.autoDispose.family<List<GuestPromo>?, String>((ref, guestId) {
  return ref.watch(guestPromosStreamProvider(guestId).future);
});
```

To use the same actions from different screens, I have my actions (things that can be triggered by button taps) done with a "top-level-provider" which gives me the ability to get a ProviderRef to an instance of a class that provides different actions.

```
final guestActionsProvider = Provider<GuestActions>((ref) {
  return GuestActions(ref);
});
```

Inside these GuestActions, I want to be able to able to read the current state of the guestPromosProvider like this:

```
class GuestActions {
  final ProviderRef ref;

  GuestActions(this.ref);
  
  Future<List<Payment>?> guestPays(Guest payingGuest) async {
      final guestPromos = await ref.read(guestPromosProvider (payingGuest.id!).future);
      ....
  }
  ....
}
```

I use this action class now from anywhere, always triggered by a button-tap or something similar with the following code:

```
 onTap: () async {
 ref.read(guestActionsProvider).guestsPay(selectedGuest);
}
```

Now, the fun thing is: this actually works - but only sometimes :-( 

Sometimes the error below appears in the console - but it works 100% the second time I press the same button. So my guess is, that there is some problem with the .future - which doesn't really wait until the stream is actually fully loaded.

Now my question is:
Is this a correct way of implementing the use case I described? 
Thank you all for your suggestions.

```
Error: Bad state: The provider AutoDisposeFutureProvider<List<GuestPromo>?>#5d055(HQED6i1NNf1uAAlsjRj7) was disposed before a value was emitted.
    Non-error `null` thrown by JS does not have stack trace.
Caught in Dart at:
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 303:20  missing
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 167:24  stackTrace
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 180:10  stackTraceForError
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 284:38            get stackTrace
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/async/async_error.dart 26:29                                     defaultStackTrace
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/async/future_impl.dart 30:33                                     completeError
packages/riverpod/src/async_value_converters.dart 400:23                                                                   <fn>
packages/riverpod/src/framework/container.dart 5:7                                                                         _runGuarded
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/core/iterable.dart 325:35                                        forEach
packages/riverpod/src/framework/provider_base.dart 909:26                                                                  [_runOnDispose]
packages/riverpod/src/framework/provider_base.dart 860:5                                                                   dispose
packages/riverpod/src/framework/container.dart 354:12                                                                      [_disposeProvider]
packages/riverpod/src/framework/scheduler.dart 81:14                                                                       [_performDispose]
packages/riverpod/src/framework/scheduler.dart 42:5                                                                        [_task]
packages/flutter_riverpod/src/framework.dart 303:12                                                                        build
packages/flutter/src/widgets/framework.dart 4754:15                                                                        performRebuild
packages/flutter/src/widgets/framework.dart 4477:5                                                                         rebuild
packages/flutter/src/widgets/framework.dart 2659:18                                                                        buildScope
packages/flutter/src/widgets/binding.dart 882:9                                                                            drawFrame
packages/flutter/src/rendering/binding.dart 363:5                                                                          [_handlePersistentFrameCallback]
packages/flutter/src/scheduler/binding.dart 1144:15                                                                        [_invokeFrameCallback]
packages/flutter/src/scheduler/binding.dart 1081:9                                                                         handleDrawFrame
packages/flutter/src/scheduler/binding.dart 995:5                                                                          [_handleDrawFrame]
C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/platform_dispatcher.dart 1011:13           invoke
C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/platform_dispatcher.dart 159:5             invokeOnDrawFrame
C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/initialization.dart 128:45                 <fn>
    at Object.createErrorWithStack (http://localhost:9090/dart_sdk.js:5080:12)
    at Function._throw (http://localhost:9090/dart_sdk.js:20337:18)
    at Function.throwWithStackTrace (http://localhost:9090/dart_sdk.js:20334:18)
    at async._AsyncCallbackEntry.new.callback (http://localhost:9090/dart_sdk.js:40851:18)
    at Object._microtaskLoop (http://localhost:9090/dart_sdk.js:40708:13)
    at _startMicrotaskLoop (http://localhost:9090/dart_sdk.js:40714:13)
    at http://localhost:9090/dart_sdk.js:36191:9

```

**Comment by rrousselGit** [comment]
> Now, I also want to be able to get the latest version of these guestPromos on a button tap - from a screen that doesn't need to load these guestPromos in advance - it only needs them after the button is pressed.
> 
> So my idea was to use a FutureProvider to get the .future of the StreamProvider to get the current result. The Future Provider looks like this:

This is unnecessary. You can obtain that directly from the streamprovider

```ref
onTap: () => ref.read(streamProvider.future);
```


____


Your problem likely stems from the autoDispose combined with you using `ref.read`
Your provider likely isn't listened anymore and got disposed. 

You could try doing:

```dart
  Future<List<Payment>?> guestPays(Guest payingGuest) async {
      final guestPromosSub = ref.listen(guestPromosProvider(payingGuest.id!).future, (prev, next) {});
      try {
        final guestPromos = await guestPromosSub.read();
        ....
      } finally {
        guestPromosSub.close();
      }
      ....
  }
```

**Reply by perret123** [reply]
Thank you very much for your fast response!
 
About the unnecessary `FutureProvider`: You are right, while writing the text, I was also thinking that I over-complicated this, thank you for the confirmation :-)

It seems I will need to update `riverpod` to 2.0.0-dev.5 to get a return value for ref.listen, correct? Because for now (1.0.3) it tells me that `ProviderRef.listen` has a return-value of `void`. 

Also just a general question: 

I use `maintainState` and `.autoDispose` in combination with the Firestore-Streams, because I want the Stream to stop listening for changes to the database when I don't use them - but I also want the latest fetched data to be still saved in the provider.

When I need the data of one of those streams in a user-interaction, I can use `ref.read(streamProvider.future)` as you say - but there is no guarantee that this command will actually give me back data - as the `StreamProvider` may be not ready (loading) or is in some kind of disposed state (as the error suggests). 

As a workaround, I was also thinking that I could create a method which will just query the Firestore database (without streaming changes) - and make an always alive `FutureProvider` for that case. But my thinking was, that in most cases, I already have a stream with this data open and querying it again would just be wasting precious "reads" in the Firestore Database, so I rather use the (mostly) already available Stream.

So I wonder if my use case could also be achieved by something like; `await ref.read(streamProvider.latest)` which will guarantee me that the provider will be "undisposed" if neccesary and just wait out the loading state to give me the desired latest value of a Stream. 

Maybe with this ref.listen(..).read() something like this is achievable (I will try soon) - but I wonder if it is also feasible/possible to do it with less syntax in a future update OR if my thoughts about this are wrong :-)

Thank you for your time!

**Reply by rrousselGit** [reply]
Something similar has been requested before. I've opened an issue for this feature request https://github.com/rrousselGit/river_pod/issues/1344

For now if you want to pause the stream when it isn't listened, you'll have to do things manually.

But the 2.0 and its `ref.onCancel`/`ref.onResume` may help

**Reply by perret123** [reply]
Actually, I don't really want to pause the stream - this is a nice feature as well, but for my case I would rather like to have a `StreamProvider` parameter similar to `.future` - but one that ensures that there is either an error (from the stream, not the provider) or a value that can be awaited. Not sure if this is feasible or if there are practical reasons against this, but yeah, just wanted to make it clear what I think I need :-) Thank you in any case.

**Reply by rrousselGit** [reply]
I'm not sure I understand then. How is what you're asking different from `.future`?

**Reply by perret123** [reply]
With `ref.read(streamProvider.future)`, it is possible I receive a `AutoDisposeFutureProvider was disposed before a value was emitted.` exception if the provider is not currently used/listened to from somewhere else. If I understood you correctly - this is by design.

But in case of a button tap, I don't care if the provider was disposed/ is listened to, I just want to be sure it will deliver me a result on the first request - this is why I use await (I don't care if it takes some seconds even, as this is not about painting frames). 
So actually, I want a parameter that forces the provider to rebuild if not already listening and wait for a value to be emmited from the stream. In best case I want to be able to do that with a autoDispose-Provider and .read - the provider can be disposed right after this again - if nobody else is listening to it. 

Maybe this is some kind of double-use that is not wanted, but it would surely reduce complexity if i could just use a StreamProvider in async-action-methods without having it used in a `build()`-method at the same time. 

Maybe to give you the gist of my use-case - it is about a restaurant app:
I have a screen which displays maybe 30 guests on different tables in a LocationView. Each of these guests have products attached to them. But I only use a ref.watch on my `guestProductsStreamProvider` when the guest is actually opened on the screen (GuestView). But now I have a context menu already inside the LocationView which enables the user to start the payment process without actually opening the GuestView and start to listen to the `guestProductsStreamProvider` - so I want to use it on a button-press right away without it being listened to before. 
I already tried to have the `guestProductsStreamProvider` listened to inside the context-menu - but as the context menu needs to be disposed (as the payment process opens other popups) this was no solution.

Funnily, it works sometimes, so maybe `.future` is already what this should be, but then maybe this is rather a bug report then a feature request?

**Reply by perret123** [reply]
Addition:
I updated to `riverpod 2.0.0-dev.5` and have the same error. But I will try to use the new methods provided to keep a provider alive and the `ref.listen` `.read` method to see if that changes things.

**Reply by perret123** [reply]
Addition 2:

I tried to use
```dart
  final guestPromosSub = ref.listen(guestPromosProvider(payingGuest.id!).future, (prev, next) {});
      try {
        final guestPromos = await guestPromosSub.read();
```
but I got the same exception. I guess in the end, this is similar to `ref.read(streamProvider.future)` ? 


Then I tried to update my provider with the new `keepAlive()` method;

```dart
final guestProductStreamProvider = StreamProvider.autoDispose.family<List<GuestProduct>, String?>((ref, guestId) {
  final database = ref.watch(databaseProvider);
  final businessId = ref.watch(businessIdProvider);
  ref.keepAlive();
  ref.onDispose(() {
    debugLog('DISPOSED: guestProductStreamProvider');
  });
  return businessId != null && guestId != null ? database.guestProductsStream(businessId, guestId) : const Stream.empty();
});
```

And I tried to use the new `ref.refresh()` method, as I had hopes this would maybe help as it forces the provider to rebuild:

`await ref.refresh(guestProductStreamProvider(guestId!).future)`

But I run into the same error again:
`[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: Bad state: The provider AutoDisposeStreamProvider<List<GuestProduct>>#05378(pGLHo6mVMPWICWxKvsBs) was disposed before a value was emitted.`
And what is also interesting: ref.onDispose was only called before the exception, not after. So ref.refresh worked in reseting the provider.

Could this maybe be a problem caused by how Streams behave?

**Reply by perret123** [reply]
Sorry for the spam of answers :-)
I have made a reproducible example with `riverpod 2.0.0-dev.5` to make my point more clear, please see below:

When you try ref.read(myStreamProvider.future) without the provider being watched, it will end in the above mentioned "The Provider was disposed before a value was emitted".

When you try the same while the provider is still loading, it will end in the same exception.

In theory I would expect that streamProvider.future would ignore the loading period and just give me the value whenever it is ready. But yeah, maybe my understanding of how Streams work is limited - but somehow I am sure this must be possible as the provider knows the value is loading, so it must somehow also know how to wait for the first result. 

What do you think?

```dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(const ProviderScope(child: const MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

final myStreamProvider = StreamProvider.autoDispose<int>((ref) {
  return Stream<int>.periodic(const Duration(seconds: 5), (x) => x).take(15);
});

class MyHomePage extends ConsumerStatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  ConsumerState<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends ConsumerState<MyHomePage> {
  void _readValueFromProvider() async {
    final result = await ref.read(myStreamProvider.future);
    print(result);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElevatedButton(onPressed: _readValueFromProvider, child: Text('Read value from Stream when the provider is not alive.')),
            ElevatedButton(
                onPressed: () {
                  showDialog(
                      context: context,
                      builder: (context) {
                        return Center(child: Card(child: Consumer(builder: (context, ref, child) {
                          var result = ref.watch(myStreamProvider);
                          return Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              result.when(
                                  data: (data) => Text(data.toString()),
                                  loading: () => CircularProgressIndicator(),
                                  error: (e, stck) => Text(e.toString())),
                              ElevatedButton(
                                  onPressed: _readValueFromProvider, child: Text('Read value from Stream when the provider is alive.'))
                            ],
                          );
                        })));
                      });
                },
                child: Text('Read value from Stream when Provider is alive.'))
          ],
        ),
      ),
    );
  }
}
```

**Reply by perret123** [reply]
@rrousselGit 

It is me again :-) 

I was checking the river_pod source code for the first time to see if there might be a solution for my problem, and I may have found at least a way to solve my specific use-case, but now the question is if this makes sense overall or will lead to other problems.

In the `ProviderElementBase` class I added 
```dart
bool? loadingCompleted;
```

This variable is initialized/set in the `_asyncValueAsFuture` method which I added to the class `AutoDisposeAsyncValueAsFutureProvider` - as there the knowledge about the current state of the loading process seems to be handled.

```dart
  Future<State> _asyncValueAsFuture(ProviderBase<AsyncValue<State>> provider,
      ProviderElementBase<Future<State>> ref) {
    Completer<State>? loadingCompleter;

    ref.onDispose(() {
      if (loadingCompleter != null) {
        ref.loadingCompleted = true;
        loadingCompleter!.completeError(
          StateError(
            'The provider $provider was disposed before a value was emitted.',
          ),
        );
      }
    });

    void listener(AsyncValue<State>? previous, AsyncValue<State> value) {
      if (value.isLoading || value.isRefreshing) {
        if (loadingCompleter == null) {
          ref.loadingCompleted = false;
          loadingCompleter = Completer<State>();
          ref.setState(
            // TODO test ignore
            loadingCompleter!.future..ignore(),
          );
        }
      }

      value.map(
        loading: (_) {
          // already taken care of above
        },
        data: (data) {
          // already taken care of above
          if (data.isRefreshing) return;

          if (loadingCompleter != null) {
            ref.loadingCompleted = true;
            loadingCompleter!.complete(data.value);
            // allow follow-up data calls to go on the 'else' branch
            loadingCompleter = null;
          } else {
            if (data.value != null) {
              ref.setState(Future<State>.value(data.value));
            }
          }
        },
        error: (error) {
          // already taken care of above
          if (error.isRefreshing) return;

          if (loadingCompleter != null) {
            ref.loadingCompleted = true;
            loadingCompleter!.completeError(error.error, error.stackTrace);
            // allow follow-up error calls to go on the 'else' branch
            loadingCompleter = null;
          } else {
            ref.setState(Future<State>.error(error.error, error.stackTrace));
          }
        },
      );
    }

    ref.listen<AsyncValue<State>>(provider, listener, fireImmediately: true);

    return ref.requireState;
  }
```

Inside the `AutoDisposeProviderElementBase` base class I added a check for the value of the new `loadingCompleted` variable to the `mayNeedDispose`-method:

```dart
  @override
  void mayNeedDispose() {
    // ignore: deprecated_member_use_from_same_package
    if (!maintainState &&
        !hasListeners &&
        _keepAliveLinks.isEmpty &&
        (loadingCompleted == null || loadingCompleted == true)) {
      _container._scheduler.scheduleProviderDispose(this);
    }
  }
```


Now, this seems to work like a charm in my app! At least my first testing session with 2 new features are working now well when before they didn't. :-) 

I ran the riverpod tests over the solution - and I have one fail in the `throws StateError if the provider is disposed before a value was emitted` test. 

So - even if seems to work for my use case, I have no bigger picture if this may messes up something in the base riverpod architecture in which I am no expert in - for me it feels like I made a dirty quickfix which may works great one day but fails another, so I would appreciate your feedback. If you want a pull request right away I can generate one, for now I was still a bit to unsure if this is really all that is needed, it feels like I am missing something.

**Reply by perret123** [reply]
Hey @rrousselGit 

Sorry for bothering and answering 10 times to my own issue. 

I was just wondering about the following things:

1.  Did you understand the problem I am facing? In short: a StreamProvider used in an async method with await ref.read/watch(streamProvider.future) will dispose before the result is delivered if the stream was not already watched/loaded at the time of use)  - see example [here](https://github.com/rrousselGit/riverpod/discussions/1343#discussioncomment-2526655 )
2. If the solution I proposed [here](https://github.com/rrousselGit/riverpod/discussions/1343#discussioncomment-2535347) makes any sense to you - because if so, I could even contribute something, but as mentioned, I am afraid it goes against some principles and I don't want to spend effort in pull-requesting a subobtimal solution.

Thank you for your time, and I understand of course that you have probably enough other requests - just wondering.

**Reply by thomas-Dooda** [reply]
Hey guys, 

I just want to join your discussion, because I am facing a similar issue. 
I am using flutter 3.0.2 (stable) and riverpod 2.0.0-dev.9

I am using a StreamProvider to provide the content for a ConsumerStateful widget, displaying a profile - The ProfileViewPage. Following is the code for the profileStreamProvider:
```
final profileStreamProvider =
    StreamProvider.autoDispose.family<Profile, String>((ref, profileId) {
  final database = ref.watch(databaseProvider);
  if (database == null) {
    print('database null.');
    return const Stream.empty();
  }
  ref.keepAlive();
  return database.profileStream(profileId: profileId);
});
```

This works fine to load the content of the according profile to the profileViewPage. Now I want to display different alert messages depending on the owner of the profile (the owner is an attribute of the profile). I am using the initState() function of the ProfileViewPageState to do so. I assume the following Future Widget to be built after the ProfileViewPageState already got the profile content from the StreamProvider.
```
@override
  void initState() {
    Future.delayed(Duration.zero, () => showProfileViewOnboarding())
        .then((userIsProfileOwner) => {
              if (userIsProfileOwner)
                displayQrcodeIfNoHandoverYet(
                  context: context,
                  ref: ref,
                  profileId: widget.profileId,
                )
            });
    super.initState();
  }
```

The showProfileViewOnboarding() is checking, who the profile owner is.
```
Future<bool> showChatViewOnboarding() async {
    final ProfilesChecker profilesChecker = ProfilesChecker(ref: ref);
    return profilesChecker.checkUserIsProfileOwner(widget.profileId);
}
```

And here I want to check, whether the owner or some other user is viewing the profile.
```
Future<bool> checkUserIsProfileOwner(String profileId) async {
    final FirebaseAuth firebaseAuth = ref.read(firebaseAuthProvider);
    final User? user = firebaseAuth.currentUser;
    if (user == null) {
      return false;
    }
    final Profile profile =
        await ref.read(profileStreamProvider(profileId).future);
    return user.uid == profile.owner;
```


But unfortunately opening the ProfileViewPage the first time, I get the following error:
```
Error: Bad state: The provider AutoDisposeStreamProvider<Profile>#d5685(oNVxkOUvom0pWZ5RxDrt) was disposed before a value was emitted.
    Non-error `null` thrown by JS does not have stack trace.
```

Interestingly, the error occures only the first time loading the page. When reloading it, it works fine.



The proposed solution using the .listen() function doesn't work for me. Using the following code in checkUserIsProfileOwner() returns the error "This expression has a type of 'void' so its value can't be used. ".
```
final profileStreamSub = ref.listen(profileStreamProvider(profileId).future, (prev, next) {});
    try {
      final profile = await profileStreamSub.read();
    } finally {
    profileStreamSub.close();
    }
```

**Reply by rivella50** [reply]
When listening to your provider you should check the next value.
And regarding your 'Bad state' error: I thought that the first value is always null (since async) and only the next event contains then the needed value.

**Comment by perret123** [comment, marked as answer]
@rrousselGit @thomas-Dooda

By trying to work around the issue I figured out a better way to manage my streams and make my application a lot more efficient.

1. Whenever you open a stream to Firestore, keep it exactly alive for as long as you need it - like this you can reduce read costs in Firestore. 
2. Even if the StreamProvider is .autoDispose, it will not close the actual StreamSubscription from Firestore! You will need to take care of that yourself by closing the subscription "manually". I was quite surprised by this - and it was quite hard to figure out as Firestore doesn't have good enough logs.

With that knowledge, I implemented my providers the following way:

Some background: I build a restaurant app and load some data connected to guests - these streams should stay alive for as long as the guest is still at some location.
So for all my guestProviders, I created a global method which creates a `keepAliveLink` which will be closed as soon as the guest is not at the location anymore (by listening to the guestLocationIdProvider). And if the guest has "left" it will close the Firestore stream with `sub?.cancel()`.

```dart
void prepareGuestStreamProvider<T>(AutoDisposeStreamProviderRef<T> ref,
    StreamSubscription<dynamic>? sub, String guestId) {
  final keepAliveLink = ref.keepAlive();
  ref.listen<String?>(guestLocationIdProvider(guestId), (previous, next) {
    if (next == null) {
      keepAliveLink.close();
    }
  }, fireImmediately: true);

  ref.onDispose(() {
    sub?.cancel();
  });
}
```
Now, I can use this method with for example my guestVouchers and guestProducts StreamProviders:

```dart

final guestProductStreamProvider = StreamProvider.autoDispose
    .family<List<GuestProduct>, String>((ref, guestId) {
  final database = ref.watch(databaseProvider);
  final businessId = ref.watch(businessIdProvider);

  if (businessId == null) return const Stream.empty();
  final stream = database.guestProductsStream(businessId, guestId);
  prepareGuestStreamProvider(
    ref,
    stream.subscription,
    guestId
  );

  return stream.result;
});


final guestVouchersStreamProvider =
    StreamProvider.autoDispose.family<List<Voucher>, String>((ref, guestId) {
  final database = ref.watch(databaseProvider);
  final businessId = ref.watch(businessIdProvider);

  if (businessId == null) return const Stream.empty();
  final stream = database.guestVouchersStream(businessId, guestId);
  prepareGuestStreamProvider(
      ref, stream.subscription, guestId);
  return stream.result;
});

```

My database methods are setup as follows:

To get not only the Stream of Results, but also the StreamSubscription to close later, I created a class that gives me both;

```dart
class SubscriptionAndResult<T> {
  StreamSubscription<dynamic>? subscription;
  Stream<T> result;
  SubscriptionAndResult({
    required this.subscription,
    required this.result,
  });
}
```

With that class, I use it to create my Firestore-Streams like this:

```dart

  SubscriptionAndResult<List<T>> collectionStream<T>({
    required String path,
    required T Function(
            Map<String, dynamic> data, String documentID, DocumentReference ref)
        builder,
    Query<Map<String, dynamic>>? Function(Query<Map<String, dynamic>> query)?
        queryBuilder,
    int Function(T lhs, T rhs)? sort,
  }) {
    Query<Map<String, dynamic>> query = fireStoreInstance.collection(path);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }

    final snapshots = query.snapshots();
    return SubscriptionAndResult(
        subscription: snapshots.listen(null),
        result: snapshots.map((snapshot) {
          final result = snapshot.docs
              .map((snapshot) {
                return builder(
                    snapshot.data(), snapshot.id, snapshot.reference);
              })
              .where((value) => value != null)
              .toList();
          if (sort != null) {
            result.sort(sort);
          }
          return result;
        }));
  }

```

And for example to get the vouchers of a specific guest, I create the stream like this:

```dart
  SubscriptionAndResult<List<Voucher>> guestVouchersStream(
          String businessId, String guestId) =>
      _service.collectionStream(
        path: FirestorePath.vouchers(businessId),
        queryBuilder: (query) =>
            isActive(query).where('attachedToGuest', isEqualTo: guestId),
        builder: (data, documentId, ref) =>
            Voucher.fromMap(data, documentId, ref),
      );
``` 

My first solution for setting up my Firestore-Connection was very similar as in the [riverpod-firestore-starter-architecture ](https://codewithandrea.com/videos/starter-architecture-flutter-firebase/) - but I ran into issues and app crashes because the streams were not closed properly or closed too early (hence my original issue with the too early closing AutoDisposeStreamProviders). 

So actually, I am quite happy now with the keepAliveLink-Solution - so my issue with the FutureProvider closing too early is actually not really relevant anymore. I am sure this solution could still be improved upon - but for now I am happy, as my Firestore reads have dramatically decreased.

I hope this helps some others using Firestore and Riverpod as well :-)

**Reply by rivella50** [reply]
Regarding the not closing Firestore subscription i have a similar problem: https://github.com/rrousselGit/riverpod/issues/1468
I will definitely try to cancel the subscription, as you're suggesting.
Thanks for your lines.

**Reply by rrousselGit** [reply]
> 2\. Even if the StreamProvider is .autoDispose, it will not close the actual StreamSubscription from Firestore!

What do you mean? It does close the subscription

Looking at your code, my guess is that you listened to the stream on your own. And _that_ subscription wasn't closed

If only StreamProvider is listening to the stream, you shouldn't have such an issue

**Reply by rivella50** [reply]
@perret123 Unfortunately it doesn't solve my problem...

**Reply by perret123** [reply]
@rrousselGit 
I am quite positive that only the StreamProvider was listening to the Stream. As it wasn't used anywhere else. But I am no expert in Stream handling, so maybe my thoughts are wrong but this is how i see it;

```dart   
final snapshots = query.snapshots();
```
This gave me a Stream of `<QuerySnapshot<Map<String, dynamic>>>` -- and then I was using `snapshots.map((snapshot) { ... }` to converse this to a  `Stream<Product>` (as an example). And the StreamProvider.autoDispose closed the `Stream<Product>` on its disposal, but I don't think the original Firestore Stream with `<QuerySnapshot<Map<String, dynamic>>>` was really closed with that ... this is why I now opted to make sure myself that the original Firestore-Stream is closed ;D I only manually tested this in the middle of the night and was quite convinced by my finding back then. But if you are sure I am wrong with this, please tell me. I would be happy to have less bloated sourcecode.

@rivella50 
Hmm as far as I understand your problem, you don't really want to cancel your subscription - as your problem seems to have to do with how firestore works with persistance enabled - at least it doesn't seem like a riverpod issue to me. But maybe I am wrong here as well :D

**Reply by rivella50** [reply]
@perret123 Yes you're right. It doesn't seem to be a problem with Riverpod (btw. i've seen the same behaviour when testing with the new framework [Creator](https://pub.dev/packages/creator)).
What i possibly could do for snapshot.docs is adding a check (e.g. with a bool isFirstSnapshot) for not handling a cached value if it's the first snapshot of a stream when (re-)listening to it. Not nice but it would work.
This won't work for snapshot.docChanges though since the follow up snapshots only contain the changes in comparison to the first (cached) value.
I have created now an [issue](https://github.com/firebase/flutterfire/issues/8961) on the Firebase GitHub page to see what they think about it.

Btw. when debugging the Riverpod code i have seen that it also cancels stream subscriptions when the provider is disposed.

**Reply by rivella50** [reply]
@perret123 But you could do me a favour and check if you're getting cached values as first notification when relistening a stream provider of yours. We have very similar code therefore it would be interesting for me.

**Reply by perret123** [reply]
@rivella50 
Omg, 5 months have passed. Sorry about that.
Yeah, I think I've got the same problem as you describe. I "manually" dispose the StreamProviders on actions such as Logout, but after logging in again, I get some "ghost" data which shouldn't even appear with the currently logged in user. 
Seems to mirror the problem description you mentioned here: https://github.com/firebase/flutterfire/issues/8961 
So it rather is a Firebase-Issue then a Riverpod Issue. If I find any way of getting around this behavior, I'll update you.

**Reply by rivella50** [reply]
@perret123 Yes, you're right, it's a Firebase issue. As you can see in the mentioned FlutterFire issue i've opened a feature request in order to introduce some sort of GetOptions also for `Query.snapshots` and `DocReference.snaphots`.
Another possibility: Checkout the package [Nimbostratus](https://pub.dev/packages/nimbostratus) which extends Cloud Firestore and introduces data fetching policies.
